straxen/plugins/peak_processing.py

```diff
@@ -332,6 +332,103 @@ def find_n_competing(peaks, windows, fraction):
+                 default=('electron_drift_velocity', 'ONLINE', True),
+                 help='Vertical electron drift velocity in cm/ns (1e4 m/ms)'),
+    strax.Option(name='max_drift_length', default=straxen.tpc_z,
+                 help='Total length of the TPC from the bottom of gate to the '
+                      'top of cathode wires [cm]'),
+    strax.Option(name='exclude_drift_time', default=False,
+                 help='Subtract max drift time to avoid peak interference in '
+                      'a single event [ns]'))
+class PeakShadow(strax.OverlapWindowPlugin):
+    """
+    This plugin can find and calculate the previous S2 shadow at peak level,
+    with time window backward and previous S2 area as options.
+    It also gives the area and position information of these previous S2s.
+    """
+
+    __version__ = '0.1.0'
+    depends_on = ('peak_basics', 'peak_positions')
+    provides = 'peak_shadow'
+    save_when = strax.SaveWhen.EXPLICIT
+
+    def setup(self):
+        self.time_window_backward = self.config['time_window_backward']
+        if self.config['exclude_drift_time']:
+            electron_drift_velocity = straxen.get_correction_from_cmt(
+                self.run_id,
+                self.config['electron_drift_velocity'])
+            drift_time_max = int(self.config['max_drift_length'] / electron_drift_velocity)
+            self.n_drift_time = drift_time_max
+        else:
+            self.n_drift_time = 0
+        self.s2_threshold = self.config['pre_s2_area_threshold']
+        self.exponent = self.config['deltatime_exponent']
+
+    def get_window_size(self):
+        return 3 * self.config['time_window_backward']
+
+    def infer_dtype(self):
+        dtype = [('shadow', np.float32, 'previous s2 shadow [PE/ns]'),
+                 ('pre_s2_area', np.float32, 'previous s2 area [PE]'),
+                 ('shadow_dt', np.int64, 'time difference to the previous s2 [ns]'),
+                 ('pre_s2_x', np.float32, 'x of previous s2 peak causing shadow [cm]'),
+                 ('pre_s2_y', np.float32, 'y of previous s2 peak causing shadow [cm]')]
+        dtype += strax.time_fields
+        return dtype
+
+    def compute(self, peaks):
+        roi_shadow = np.zeros(len(peaks), dtype=strax.time_fields)
+        roi_shadow['time'] = peaks['center_time'] - self.time_window_backward
+        roi_shadow['endtime'] = peaks['center_time'] - self.n_drift_time
+
+        mask_pre_s2 = peaks['area'] > self.s2_threshold
+        mask_pre_s2 &= peaks['type'] == 2
+        split_peaks = strax.touching_windows(peaks[mask_pre_s2], roi_shadow)
+        res = np.zeros(len(peaks), self.dtype)
+        res['pre_s2_x'] = np.nan
+        res['pre_s2_y'] = np.nan
+        if len(peaks):
+            self.compute_shadow(peaks, peaks[mask_pre_s2], split_peaks, self.exponent, res)
+
+        res['time'] = peaks['time']
+        res['endtime'] = strax.endtime(peaks)
+        return res
+
+    @staticmethod
+    @numba.njit
```

Comment:
Any reason why we are not caching this guy?

Reply:
It did not speed it up and I think we should only cache if there is a good reason to do so. It can create issues when you are running many jobs at once.

These are the times for computing the plugin:

With caching:
```2.79 s ± 28.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)```

without caching:
```2.73 s ± 32.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)```

Reply:
Agree